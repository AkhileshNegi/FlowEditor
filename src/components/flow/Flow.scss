@import '~/variables.scss';

.loading {
    visibility: hidden;
    height: inherit;
}

.loaded {
    height: inherit;
    position: relative;
}

.nodeList {
    position: relative;
    user-select: none;
}

.dragSelection {
    position: absolute;
    background: red;
    border: 0px solid darken($selection_square, 10%);
    opacity: 0;
    pointer-events: none;
    z-index: $z_selection;
}
