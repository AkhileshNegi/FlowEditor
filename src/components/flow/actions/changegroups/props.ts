import { ActionFormHelper, Type } from '~/config/typeConfigs';
import { ChangeGroups } from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export interface ChangeGroupsFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    formHelper: ActionFormHelper;
    typeConfig: Type;

    // update handlers
    updateAction(action: ChangeGroups): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}
export default ChangeGroupsFormProps;
