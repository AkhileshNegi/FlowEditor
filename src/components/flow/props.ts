import { UpdateLocalizations } from '~/components/nodeeditor/NodeEditor';
import { Type } from '~/config/typeConfigs';
import { AnyAction } from '~/flowTypes';
import { Asset } from '~/services/AssetService';
import { AssetStore, RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';
import { DispatchWithState, GetState } from '~/store/thunks';

export interface ActionFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;
    assets: AssetStore;

    // update handlers
    updateAction(
        action: AnyAction,
        onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
    ): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface RouterFormProps {
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;

    assets: AssetStore;

    // update handlers
    updateRouter(renderNode: RenderNode): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface LocalizationFormProps {
    language: Asset;
    nodeSettings: NodeEditorSettings;
    updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
    onClose(canceled: boolean): void;
}
