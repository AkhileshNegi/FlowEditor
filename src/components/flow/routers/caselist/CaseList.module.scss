@import 'variables.module.scss';

.case_list {
  margin: 4px -15px;
  position: relative;
  overflow-y: scroll;
  max-height: 250px;
  padding-bottom: 10px;
  padding-top: 6px;
}

.kase {
  z-index: 2000;
}
