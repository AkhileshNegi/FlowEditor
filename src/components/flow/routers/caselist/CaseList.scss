.case-list {
    margin: 0 -15px;
}