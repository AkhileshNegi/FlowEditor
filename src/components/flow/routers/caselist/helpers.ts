import { DraggableStyle } from 'react-beautiful-dnd';
import { CaseProps, DragCursor } from '~/components/flow/routers/caselist/CaseList';
import { Operators } from '~/config/operatorConfigs';
import { Case, Exit } from '~/flowTypes';
import { createUUID } from '~/utils';

export const getExitName = (kase: Case, exits: Exit[]) => {
    const match = exits.find((exit: Exit) => exit.uuid === kase.exit_uuid);
    if (match) {
        return match.name;
    }
    return '';
};

export const createEmptyCase = (): CaseProps => {
    const uuid = createUUID();
    return {
        uuid,
        kase: {
            uuid,
            type: Operators.has_any_word,
            arguments: [''],
            exit_uuid: null
        },
        exitName: ''
    };
};

export const getItemStyle = (
    draggableStyle: DraggableStyle,
    isDragging: boolean
): DraggableStyle => ({
    userSelect: 'none',
    outline: 'none',
    ...draggableStyle,
    // Overwriting default draggableStyle object from this point down
    ...(isDragging
        ? {
              background: '#f2f9fc',
              borderRadius: 4,
              opacity: 0.75,
              height: draggableStyle.height + 15
          }
        : {})
});

export const getListStyle = (isDraggingOver: boolean, single: boolean): { cursor: DragCursor } => {
    if (single) {
        return null;
    }

    return {
        cursor: isDraggingOver ? DragCursor.move : DragCursor.pointer
    };
};
