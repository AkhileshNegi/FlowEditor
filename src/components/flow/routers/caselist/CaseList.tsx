import arrayMove from 'array-move';
import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { SortableContainer, SortableElement, SortEnd } from 'react-sortable-hoc';
import CaseElement from '~/components/flow/routers/case/CaseElement';
import { createEmptyCase } from '~/components/flow/routers/caselist/helpers';
import { fakePropType } from '~/config/ConfigProvider';
import { Case } from '~/flowTypes';
import { FormState, mergeForm } from '~/store/nodeEditor';

import * as styles from './CaseList.scss';

export enum DragCursor {
    move = 'move',
    pointer = 'pointer'
}

export interface CaseProps {
    uuid: string;
    kase: Case;
    categoryName: string;
    valid: boolean;
}

export interface CaseListProps {
    cases: CaseProps[];
    onCasesUpdated(cases: CaseProps[]): void;
}

export interface CaseListState extends FormState {
    currentCases: CaseProps[];
}

/**
 * CaseList is a component made up of case elements that lets
 * the user configure rules and drag and drop to set their order.
 */
export default class CaseList extends React.Component<CaseListProps, CaseListState> {
    private sortableItem = SortableElement(({ value: caseProps }: any) => (
        <div className={styles.kase}>
            <CaseElement
                key={caseProps.uuid}
                {...caseProps}
                onRemove={this.handleRemoveCase}
                onChange={this.handleUpdateCase}
            />
        </div>
    ));

    private sortableList = SortableContainer(({ items }: any) => {
        return (
            <div className={styles.caseList}>
                {items.map((value: any, index: any) => (
                    <this.sortableItem
                        key={`item-${index}`}
                        index={index}
                        value={value}
                        disabled={index === this.state.currentCases.length - 1}
                        shouldCancelStart={(e: any) => {
                            console.log(e);
                            return true;
                        }}
                    />
                ))}
            </div>
        );
    });

    constructor(props: CaseListProps) {
        super(props);

        bindCallbacks(this, {
            include: [/^handle/]
        });

        const caseProps = this.props.cases;

        if (!this.hasEmptyCase(caseProps)) {
            caseProps.push(createEmptyCase());
        }

        // initialize our cases
        this.state = {
            currentCases: caseProps,
            valid: true
        };
    }

    public static contextTypes = {
        config: fakePropType
    };

    private handleUpdate(keys: { caseProps?: CaseProps; removeCase?: any }): boolean {
        const updates: Partial<CaseListState> = {};

        let ensureEmptyCase = false;

        if (keys.hasOwnProperty('caseProps')) {
            updates.currentCases = [keys.caseProps];
            ensureEmptyCase = true;
            if (!keys.caseProps.valid) {
                updates.valid = false;
            }
        }

        let toRemove: any[] = [];
        if (keys.hasOwnProperty('removeCase')) {
            toRemove = [{ currentCases: [keys.removeCase] }];
            ensureEmptyCase = true;
        }

        const updated = mergeForm(this.state, updates, toRemove) as CaseListState;

        // notify our listener
        this.props.onCasesUpdated(updated.currentCases);

        // update our form
        this.setState(updated, () => {
            // if we no longer have an empty case, add one
            if (ensureEmptyCase) {
                if (!this.hasEmptyCase(this.state.currentCases)) {
                    this.handleUpdate({ caseProps: createEmptyCase() });
                }
            }
        });
        return updated.valid;
    }

    private hasEmptyCase(cases: CaseProps[]): boolean {
        return (
            cases.find((caseProps: CaseProps) => caseProps.categoryName.trim().length === 0) != null
        );
    }

    private handleRemoveCase(uuid: string): boolean {
        return this.handleUpdate({
            removeCase: { uuid }
        });
    }

    private handleUpdateCase(caseProps: CaseProps): boolean {
        return this.handleUpdate({ caseProps });
    }

    private handleSortEnd({ oldIndex, newIndex }: SortEnd): void {
        this.setState(
            ({ currentCases }) => ({
                currentCases: arrayMove(
                    currentCases,
                    oldIndex,
                    newIndex === this.state.currentCases.length - 1 ? newIndex - 1 : newIndex
                )
            }),
            () => {
                this.props.onCasesUpdated(this.state.currentCases);
            }
        );
    }

    public render(): JSX.Element {
        return (
            <>
                <this.sortableList
                    items={this.state.currentCases}
                    onSortEnd={this.handleSortEnd}
                    shouldCancelStart={(e: React.MouseEvent<HTMLDivElement>) => {
                        if (!(e.target instanceof HTMLElement)) {
                            return true;
                        }
                        return !e.target.dataset.draggable;
                    }}
                />
            </>
        );
    }
}
