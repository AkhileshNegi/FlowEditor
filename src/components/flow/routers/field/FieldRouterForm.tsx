import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet } from '~/components/dialog/Dialog';
import { sortFieldsAndProperties } from '~/components/flow/actions/updatecontact/helpers';
import { CONTACT_PROPERTIES } from '~/components/flow/actions/updatecontact/UpdateContactForm';
import { RouterFormProps } from '~/components/flow/props';
import CaseList, { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import AssetSelector from '~/components/form/assetselector/AssetSelector';
import OptionalTextInput from '~/components/form/optionaltext/OptionalTextInput';
import TypeList from '~/components/nodeeditor/TypeList';
import { fakePropType } from '~/config/ConfigProvider';
import { Scheme, SCHEMES } from '~/config/typeConfigs';
import { Asset, AssetType } from '~/store/flowContext';
import { AssetEntry, FormState, StringEntry } from '~/store/nodeEditor';
import { small } from '~/utils/reactselect';

import * as styles from './FieldRouterForm.scss';
import { nodeToState, stateToNode } from './helpers';

// TODO: Remove use of Function
// tslint:disable:ban-types
export enum InputToFocus {
    args = 'args',
    min = 'min',
    max = 'max',
    exit = 'exit'
}

export interface FieldRouterFormState extends FormState {
    field: AssetEntry;
    cases: CaseProps[];
    resultName: StringEntry;
}

export const leadInSpecId = 'lead-in';

export default class FieldRouterForm extends React.Component<
    RouterFormProps,
    FieldRouterFormState
> {
    private ROUTABLE_FIELDS: Asset[] = [
        ...CONTACT_PROPERTIES,
        ...SCHEMES.map((scheme: Scheme) => ({
            name: scheme.name,
            id: scheme.scheme,
            type: AssetType.Scheme
        }))
    ];

    constructor(props: RouterFormProps) {
        super(props);

        this.state = nodeToState(this.props.nodeSettings, this.props.assetStore);

        bindCallbacks(this, {
            include: [/^on/, /^handle/]
        });
    }

    public static contextTypes = {
        assetService: fakePropType
    };

    private handleUpdateResultName(value: string): void {
        this.setState({ resultName: { value } });
    }

    private handleFieldChanged(selected: Asset[]): void {
        this.setState({ field: { value: selected[0] } });
    }

    private handleCasesUpdated(cases: CaseProps[]): void {
        this.setState({ cases });
    }

    private handleSave(): void {
        if (this.state.valid) {
            this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
            this.props.onClose(false);
        }
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    public renderEdit(): JSX.Element {
        const typeConfig = this.props.typeConfig;

        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />
                <div className={styles.leadIn}>
                    If the contact's
                    <div className={`${styles.fieldSelect} select-small`}>
                        <AssetSelector
                            name="Contact Field"
                            styles={small}
                            assets={this.props.assetStore.fields}
                            additionalOptions={this.ROUTABLE_FIELDS}
                            entry={this.state.field}
                            searchable={true}
                            sortFunction={sortFieldsAndProperties}
                            onChange={this.handleFieldChanged}
                        />
                    </div>
                </div>
                <CaseList
                    data-spec="cases"
                    cases={this.state.cases}
                    onCasesUpdated={this.handleCasesUpdated}
                />
                <OptionalTextInput
                    name="Result Name"
                    value={this.state.resultName}
                    onChange={this.handleUpdateResultName}
                    toggleText="Save as.."
                    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
                />
            </Dialog>
        );
    }

    public render(): JSX.Element {
        return this.renderEdit();
    }
}
