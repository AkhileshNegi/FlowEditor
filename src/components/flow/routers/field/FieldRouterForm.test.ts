import { RouterFormProps } from '~/components/flow/props';
import { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import { DEFAULT_OPERAND } from '~/components/nodeeditor/constants';
import { Operators } from '~/config/operatorConfigs';
import { Types } from '~/config/typeConfigs';
import { RouterTypes, SwitchRouter } from '~/flowTypes';
import { AssetType } from '~/store/flowContext';
import { composeComponentTestUtils, mock } from '~/testUtils';
import { createRenderNode, getRouterFormProps } from '~/testUtils/assetCreators';
import * as utils from '~/utils';

import FieldRouterForm from './FieldRouterForm';

const { setup } = composeComponentTestUtils<RouterFormProps>(
    FieldRouterForm,
    getRouterFormProps(
        createRenderNode({
            actions: [],
            exits: [],
            ui: {
                position: { left: 0, top: 0 },
                type: Types.split_by_contact_field,
                config: {
                    id: 'favorite_color',
                    type: AssetType.Field
                }
            }
        })
    )
);

mock(utils, 'createUUID', utils.seededUUIDs());

const exitOneUUID = utils.createUUID();
const redCaseUUID = utils.createUUID();

describe(FieldRouterForm.name, () => {
    it('should render', () => {
        const { wrapper } = setup(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('initializes', () => {
        const { wrapper } = setup(true, {
            nodeSettings: {
                $set: {
                    originalNode: createRenderNode({
                        actions: [],
                        exits: [{ destination_node_uuid: null, name: 'Other', uuid: exitOneUUID }],
                        router: {
                            type: RouterTypes.switch,
                            operand: DEFAULT_OPERAND,
                            cases: [
                                {
                                    uuid: redCaseUUID,
                                    type: Operators.has_any_word,
                                    arguments: ['red'],
                                    exit_uuid: null
                                }
                            ],
                            default_exit_uuid: exitOneUUID,
                            result_name: 'Color'
                        } as SwitchRouter,
                        ui: {
                            position: { left: 0, top: 0 },
                            type: Types.split_by_contact_field,
                            config: {
                                id: 'favorite_color',
                                type: AssetType.Field
                            }
                        }
                    })
                }
            }
        });

        expect(wrapper).toMatchSnapshot();
    });

    describe('updates', () => {
        it('should save changes', () => {
            const { instance, props } = setup(true, {
                $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
            });

            instance.handleUpdateResultName('Favorite Color');
            instance.handleCasesUpdated([
                { kase: { type: Operators.has_any_word, arguments: ['red'] }, exitName: 'Red' },
                { kase: { type: Operators.has_any_word, arguments: ['maroon'] }, exitName: 'Red' },
                { kase: { type: Operators.has_any_word, arguments: ['green'] }, exitName: 'Green' }
            ] as CaseProps[]);

            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).toHaveBeenCalled();
            expect(props.updateRouter).toMatchCallSnapshot();
        });

        it('should cancel', () => {
            const { instance, props } = setup(true, {
                $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
            });

            instance.handleFieldChanged({ id: 'viber', name: 'Viber', type: AssetType.URN });
            instance.getButtons().secondary.onClick();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).not.toHaveBeenCalled();
        });
    });
});
