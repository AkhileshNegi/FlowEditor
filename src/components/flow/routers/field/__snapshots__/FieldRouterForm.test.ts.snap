// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldRouterForm initializes 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "disabled": false,
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_contact_field"
  title="Split by Contact Field"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by a contact field",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Contact Field",
        "type": "split_by_contact_field",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    If the contact's
    <div
      className="fieldSelect select-small"
    >
      <AssetSelector
        additionalOptions={
          Array [
            Object {
              "id": "name",
              "name": "Name",
              "type": "property",
            },
            Object {
              "id": "language",
              "name": "Language",
              "type": "property",
            },
            Object {
              "id": "channel",
              "name": "Channel",
              "type": "property",
            },
            Object {
              "id": "ext",
              "name": "External ID",
              "type": "scheme",
            },
            Object {
              "id": "facebook",
              "name": "Facebook ID",
              "type": "scheme",
            },
            Object {
              "id": "fcm",
              "name": "Firebase ID",
              "type": "scheme",
            },
            Object {
              "id": "jiochat",
              "name": "Jiochat ID",
              "type": "scheme",
            },
            Object {
              "id": "line",
              "name": "Line ID",
              "type": "scheme",
            },
            Object {
              "id": "mailto",
              "name": "Email Address",
              "type": "scheme",
            },
            Object {
              "id": "tel",
              "name": "Phone Number",
              "type": "scheme",
            },
            Object {
              "id": "telegram",
              "name": "Telegram ID",
              "type": "scheme",
            },
            Object {
              "id": "twitter",
              "name": "Twitter Handle",
              "type": "scheme",
            },
            Object {
              "id": "twitterid",
              "name": "Twitter ID",
              "type": "scheme",
            },
            Object {
              "id": "wechat",
              "name": "Wechat ID",
              "type": "scheme",
            },
            Object {
              "id": "whatsapp",
              "name": "Whatsapp Number",
              "type": "scheme",
            },
            Object {
              "id": "viber",
              "name": "Viber ID",
              "type": "scheme",
            },
          ]
        }
        entry={
          Object {
            "value": Object {
              "id": "favorite_color",
              "name": null,
              "type": "field",
            },
          }
        }
        name="Contact Field"
        onChange={[Function]}
        searchable={true}
        sortFunction={[Function]}
        styles={
          Object {
            "control": [Function],
            "indicatorsContainer": [Function],
            "input": [Function],
            "menuList": [Function],
            "option": [Function],
            "singleValue": [Function],
            "valueContainer": [Function],
          }
        }
      />
    </div>
  </div>
  <CaseList
    cases={
      Array [
        Object {
          "exitName": null,
          "kase": Object {
            "arguments": Array [
              "red",
            ],
            "exit_uuid": null,
            "type": "has_any_word",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Color",
      }
    }
  />
</Dialog>
`;

exports[`FieldRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "disabled": false,
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_contact_field"
  title="Split by Contact Field"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by a contact field",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Contact Field",
        "type": "split_by_contact_field",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    If the contact's
    <div
      className="fieldSelect select-small"
    >
      <AssetSelector
        additionalOptions={
          Array [
            Object {
              "id": "name",
              "name": "Name",
              "type": "property",
            },
            Object {
              "id": "language",
              "name": "Language",
              "type": "property",
            },
            Object {
              "id": "channel",
              "name": "Channel",
              "type": "property",
            },
            Object {
              "id": "ext",
              "name": "External ID",
              "type": "scheme",
            },
            Object {
              "id": "facebook",
              "name": "Facebook ID",
              "type": "scheme",
            },
            Object {
              "id": "fcm",
              "name": "Firebase ID",
              "type": "scheme",
            },
            Object {
              "id": "jiochat",
              "name": "Jiochat ID",
              "type": "scheme",
            },
            Object {
              "id": "line",
              "name": "Line ID",
              "type": "scheme",
            },
            Object {
              "id": "mailto",
              "name": "Email Address",
              "type": "scheme",
            },
            Object {
              "id": "tel",
              "name": "Phone Number",
              "type": "scheme",
            },
            Object {
              "id": "telegram",
              "name": "Telegram ID",
              "type": "scheme",
            },
            Object {
              "id": "twitter",
              "name": "Twitter Handle",
              "type": "scheme",
            },
            Object {
              "id": "twitterid",
              "name": "Twitter ID",
              "type": "scheme",
            },
            Object {
              "id": "wechat",
              "name": "Wechat ID",
              "type": "scheme",
            },
            Object {
              "id": "whatsapp",
              "name": "Whatsapp Number",
              "type": "scheme",
            },
            Object {
              "id": "viber",
              "name": "Viber ID",
              "type": "scheme",
            },
          ]
        }
        entry={
          Object {
            "value": Object {
              "id": "favorite_color",
              "name": null,
              "type": "field",
            },
          }
        }
        name="Contact Field"
        onChange={[Function]}
        searchable={true}
        sortFunction={[Function]}
        styles={
          Object {
            "control": [Function],
            "indicatorsContainer": [Function],
            "input": [Function],
            "menuList": [Function],
            "option": [Function],
            "singleValue": [Function],
            "valueContainer": [Function],
          }
        }
      />
    </div>
  </div>
  <CaseList
    cases={Array []}
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`FieldRouterForm updates should save changes 1`] = `
Object {
  "cases": Array [
    Object {
      "exitName": "Red",
      "kase": Object {
        "arguments": Array [
          "red",
        ],
        "type": "has_any_word",
      },
    },
    Object {
      "exitName": "Red",
      "kase": Object {
        "arguments": Array [
          "maroon",
        ],
        "type": "has_any_word",
      },
    },
    Object {
      "exitName": "Green",
      "kase": Object {
        "arguments": Array [
          "green",
        ],
        "type": "has_any_word",
      },
    },
  ],
  "field": Object {
    "value": Object {
      "id": "favorite_color",
      "name": null,
      "type": "field",
    },
  },
  "resultName": Object {
    "value": "Favorite Color",
  },
  "valid": true,
}
`;

exports[`FieldRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_node_uuid": null,
          "name": "Red",
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
        Object {
          "destination_node_uuid": null,
          "name": "Green",
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
        Object {
          "destination_node_uuid": null,
          "name": "Other",
          "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "red",
            ],
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "maroon",
            ],
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "green",
            ],
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_any_word",
          },
        ],
        "default_exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@contact.fields.favorite_color",
        "result_name": "Favorite Color",
        "type": "switch",
      },
      "uuid": "48e0a64d-3b3c-4e3e-9d95-7844093edc90",
      "wait": null,
    },
    "ui": Object {
      "config": Object {
        "operand": Object {
          "id": "favorite_color",
          "name": null,
          "type": "field",
        },
      },
      "position": null,
      "type": "split_by_contact_field",
    },
  },
]
`;
